### PR TITLE
gpt-auto: expand the loader partition UUID check to include XBOOTLDR

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 1]
```diff
@@ -727,13 +714,48 @@ static int enumerate_partitions(dev_t devnum) {
...
+                sd_id128_t loader_uuid;
+
+                r = efi_loader_get_device_part_uuid(&loader_uuid);
+                if (r < 0) {
+                        /* Won't be able to check if this is the boot ESP, thus not mounting it */
+                        esp_state = ESP_DONT_MOUNT;
+                        if (r == -ENOENT)
+                                log_debug("EFI loader partition unknown.");
+                        else
+                                log_error_errno(r, "Failed to read loader partition UUID: %m");
```
Comment (yuwata) [inline src/gpt-auto-generator/gpt-auto-generator.c]: The error will be ignored now. So, please downgrade the log level. I think, `log_debug_errno()` is sufficient. At most, `log_info_errno()`.
Then, append ", ignoring" in the message.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]
LGTM. But I'd like to wait for @poettering's final review.

**Comment by YHNdnzj** [comment]
Just rebased to resolve merge conflicts. No functional change :)

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 2]
```diff
@@ -827,7 +839,10 @@ static int enumerate_partitions(dev_t devnum) {
                         r = k;
         }
 
-        if (m->partitions[PARTITION_ESP].found) {
+        if (m->partitions[PARTITION_ESP].found &&
+            esp_is_valid(m->partitions + PARTITION_ESP,
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: indentation to 4 space is a bit weird, usually we'd indent this deeper (16spaces?), but doesn't matter

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 3]
```diff
@@ -771,6 +751,38 @@ static int add_root_mount(void) {
...
+        if (!is_efi_boot()) {
+                log_debug("Not an EFI boot, skipping ESP check.");
+                return true;
+        }
+
+        r = efi_loader_get_device_part_uuid(&loader_uuid);
+        if (r < 0) {
+                /* Won't be able to check if this is the boot ESP, thus not mounting it */
+                if (r == -ENOENT)
+                        log_debug("EFI loader partition unknown, not mounting ESP.");
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: please still use log_debug_errno() here, and pass `r` here, even if you don't use `%m` in the format string. the error number is still included in the structure log message as a separate field then, which is a good thing.

i.e. log_xyz_errno() should be used primarily when there's an error code we are acting on, and only secondarily to make %m output something nice

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 4]
```diff
@@ -771,6 +751,38 @@ static int add_root_mount(void) {
...
 
+static bool esp_is_valid(DissectedPartition *p, sd_id128_t xbootldr_uuid) {
+        sd_id128_t loader_uuid;
+        int r;
+
+        if (!is_efi_boot()) {
+                log_debug("Not an EFI boot, skipping ESP check.");
+                return true;
+        }
+
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c:765]: could you add a comment somewhere here:

```
/* Let's check if this is actually the ESP that was used for booting. If so, great. If not, let's see if the XBOOTLDR partition we found was used for booting. If so, let's use the ESP from the same disk too. If not, don't mount it */

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 5]
```diff
@@ -827,7 +839,10 @@ static int enumerate_partitions(dev_t devnum) {
                         r = k;
         }
 
-        if (m->partitions[PARTITION_ESP].found) {
+        if (m->partitions[PARTITION_ESP].found &&
+            esp_is_valid(m->partitions + PARTITION_ESP,
+                         m->partitions[PARTITION_XBOOTLDR].uuid)) {
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: this really should check `m->partitions[PARTITION_XBOOTLDR].found` before using `m->partitions[PARTITION_XBOOTLDR].uuid`

**Review by poettering — CHANGES_REQUESTED** [review]
so this looks great, some comments though.

**Comment by YHNdnzj** [comment]
Thank you for the review :)
I’m currently on vacation though. Will update when I get back.

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 6]
```diff
@@ -771,6 +751,59 @@ static int add_root_mount(void) {
...
+        /* Let's check if LoaderDevicePartUUID points to either ESP or XBOOTLDR. We prefer it pointing
+         * to the ESP, but we accept XBOOTLDR too. If it points to neither of them, don't mount any
+         * loader partitions, since they are not the ones used for booting. */
+
+        r = efi_loader_get_device_part_uuid(&loader_uuid);
+        if (r < 0) {
+                if (r == -ENOENT) {
+                        log_debug_errno(r, "EFI loader partition unknown, skipping ESP and XBOOTLDR mounts.");
+                        return 0;
+                }
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: doesn't matter, but we usually prefer shallow indentation, hence would usually write this as :

```c
if (r == -ENOENT) {
        …
}
if (r < 0) {
        …
}
`` 

but doesn't matter at all

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 7]
```diff
@@ -771,6 +751,59 @@ static int add_root_mount(void) {
...
+
+                return log_debug_errno(r, "Failed to read loader partition UUID, ignoring: %m");
+        }
+
+        if (esp->found && sd_id128_equal(esp->uuid, loader_uuid))
+                goto mount;
+
+        if (xbootldr->found && sd_id128_equal(xbootldr->uuid, loader_uuid)) {
+                log_debug("LoaderDevicePartUUID points to XBOOTLDR partition, proceeding anyway.");
```
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: "anyway" suggests this was a problem of some sort. I think the message generally makes sense for informational reasons, but I'd drop the ", proceeding anyway".

**Review by poettering — APPROVED** [review]
looks good, just some minor things that really don't matter

**Pull request merged**